src/server/fetch.ts

```diff
@@ -0,0 +1,132 @@
+/**
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { HttpsProxyAgent } from 'https-proxy-agent';
+import nodeFetch from 'node-fetch';
+import * as url from 'url';
+import { BrowserContext } from './browserContext';
+import * as types from './types';
+
+export async function playwrightFetch(context: BrowserContext, params: types.FetchOptions): Promise<{fetchResponse?: types.FetchResponse, error?: string}> {
+  try {
+    const cookies = await context.cookies(params.url);
+    const valueArray = cookies.map(c => `${c.name}=${c.value}`);
+    const clientCookie = params.headers?.['cookie'];
+    if (clientCookie)
+      valueArray.unshift(clientCookie);
+    const cookieHeader = valueArray.join('; ');
+    if (cookieHeader) {
+      if (!params.headers)
+        params.headers = {};
+      params.headers['cookie'] = cookieHeader;
+    }
+    if (!params.method)
+      params.method = 'GET';
+    let agent;
+    if (context._options.proxy) {
+      // TODO: support bypass proxy
+      const proxyOpts = url.parse(context._options.proxy.server);
+      if (context._options.proxy.username)
+        proxyOpts.auth = `${context._options.proxy.username}:${context._options.proxy.password || ''}`;
+      agent = new HttpsProxyAgent(proxyOpts);
+    }
+
+    // TODO: set user agent
+    const response = await nodeFetch(params.url, {
+      method: params.method,
+      headers: params.headers,
+      body: params.postData,
+      agent
+    });
+    const body = await response.buffer();
+    const setCookies = response.headers.raw()['set-cookie'];
+    if (setCookies) {
+      const cookies: types.SetNetworkCookieParam[] = [];
+      for (const header of setCookies) {
+        // Decode cookie value?
+        const cookie: types.SetNetworkCookieParam | null = parseCookie(header);
+        if (!cookie)
+          continue;
+        if (!cookie.domain)
```

Comment:
The same should be done to `path`. It can be omitted.

You should also perform basic same origin checks here, otherwise server can poison your cookies.

Reply:
Done.